client/components/AdminPanel.jsx

```diff
@@ -0,0 +1,165 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+
+
+import UserTile from './UserTile';
+import UserModal from './UserModal/UserModal';
+import Pagination from './Pagination/Pagination';
+import { fetchUsers, adminCreateUser, updateUser, deleteUser } from '../actions/actions';
+
+class AdminPanel extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            showModal: false,
+            isCreatingNewUser: false,
+            selectedUserToUpdate: {},
+            filterString: ''
+        };
+
+        this.fetchPage = this.fetchPage.bind(this);
+        this.openUpdateModal = this.openUpdateModal.bind(this);
+        this.onSubmitModal = this.onSubmitModal.bind(this);
+        this.submitCreateNewUser = this.submitCreateNewUser.bind(this);
+        this.submitUpdateUser = this.submitUpdateUser.bind(this);
+        this.closeModal = this.closeModal.bind(this);
+        this.deleteUser = this.deleteUser.bind(this);
+    }
+
+    componentDidMount() {
+        // Fetch the first page of users when the component mounts
+        this.props.fetchUsers(1, null);
+    }
+
+    fetchPage(page) {
+        const filter = this.state.filterString !== '' ? this.state.filterString : null;
+        this.props.fetchUsers(page, filter);
+    }
+
+    openUpdateModal(user) {
+        this.setState({
+            showModal: true,
+            selectedUserToUpdate: user
+        });
+    }
+
+    onSubmitModal(user) {
+        if (this.state.isCreatingNewUser) {
+            this.submitCreateNewUser(user);
+        } else {
+            this.submitUpdateUser(user);
+        }
+    }
+
+    async submitCreateNewUser(user) {
+        const { email, password, isAdmin } = user;
+        await this.props.adminCreateUser(email, password, isAdmin);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+
+        this.setState({
+            isCreatingNewUser: false
+        });
+
+        this.closeModal();
+    }
+
+    async submitUpdateUser(user) {
+        const { email, password, isAdmin } = user;
+        const userId = this.state.selectedUserToUpdate._id;
+        await this.props.updateUser(userId, email, password, isAdmin);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+
+        this.closeModal();
+    }
+
+    closeModal() {
+        this.setState({
+            showModal: false,
+            selectedUserToUpdate: {}
+        });
+    }
```

Comment:
these functions are `async` to make sure that one action is completed before starting the next